### ISSUE TITLE
Open transactions spike - suspected discrepancy between the metric vs the actual number of open transactions

### ISSUE THREAD

**Issue by jocrl** [cockroachdb/cockroach#76404]
# Open transactions spike

## Summary
When a customer upgraded from 21.1.13 to 21.2.5, the metric for the number of open transactions spiked to about 6k. We have reasons described below suggesting that the actual number of open transactions did not spike, but instead that the _metric_ for the number of open transactions deviated from the true count. The issue eventually went away after the customer finalized the upgrade.

Screenshots, the debug.zip, and key points from the slack thread can be found in the support ticket [here](https://github.com/cockroachlabs/support/issues/1426)

## Source of the metric
The metrics page graph is supposed to pull from the time series database, which we populate with the value over time of an atomic integer `SQLTxnsOpen`. The increment and decrement of that integer seems really simple; it is incremented in `recordTransactionStart` and decremented in `recordTransactionFinish`, which are called in the state transitions for a transaction (in `txnStateTransitionsApplyWrapper`).

## Reasons why we think it's a counter error, and not an actual spike in the number of open transactions

During the customer issue, the metrics graph showed a spike in open transactions, and so did the `"sql.txns.open"` count in the per-node `status.json` in the `debug.zip`. However, the `crdb_internal.cluster_transactions` dump in `debug.zip` only had 16 rows. And the `crdb_internal.cluster_sessions` dump only had 189 rows. (The presence of a `delegateShowTransactions` function that filters out internal transactions after querying the `crdb_internal.cluster_transactions` table suggests that the table itself does not filter out internal transactions). Additionally, even though the connections graph showed <1k connections in total across the cluster and the statements graph showed <100 active statements. Moreover, the counts for transactions that were begun, committed, and rolled back account for all transactions, as follows from the `status.json` for a node: 
```
    "sql.txn.begin.count": 227101,
    "sql.txn.commit.count": 155508,
    "sql.txn.rollback.count": 71589,
    // 155508 + 71589 = 227097 ~= 227101
```

## Failed to reproduce
In attempt to reproduce the spike, Archer started a roachprod cluster with 21.1 running TPCC, and upgraded to 21.2.5 without finalizing the upgrades. This did not reproduce a spike, nor did running kv95.

## Possible causes
There has been much thinking over what possible areas of the code could potentially cause a mis-accounting of the `SQLTxnsOpen` counter, to little avail. Details of some possible explanations explored can be found in [this slack thread](https://cockroachlabs.slack.com/archives/C0159JK877C/p1644343995543749), with key points in the support ticket linked above.

Jira issue: CRDB-13106

**Comment by MoinTom** [comment]
I encountered the same issue after upgrading from v21.2.4 to v21.2.5
![crdb_sql_txns_open](https://user-images.githubusercontent.com/3630717/153623171-2e5c6131-f319-4d59-bff8-779c3008f3b4.png)

I will be happy to assist you with any further informations.

**Comment by jocrl** [comment]
@Azhng observed that if [this](https://github.com/cockroachdb/cockroach/blob/2826900bf9e611662d9c73102c2d129fcbdb9085/pkg/sql/conn_executor_exec.go#L2175) if block is entered, the function returns without decrementing in [this line](https://github.com/cockroachdb/cockroach/blob/2826900bf9e611662d9c73102c2d129fcbdb9085/pkg/sql/conn_executor_exec.go#L2192) further down.

A fix to this should be backported to 21.2 (and anything else affected). After applying the fix, a reboot is needed to reset the counter.

**Comment by Azhng** [comment]
Fixed by https://github.com/cockroachdb/cockroach/pull/77236 and https://github.com/cockroachdb/cockroach/pull/77237
